internal/controller/streams.go

```diff
@@ -197,105 +164,225 @@ func (sm *streamManager) BootUpStreamAndReconnect(ctx context.Context, cancel co
+	}
+
+	return nil
+}
+
+// connectAndStreamResources creates resourceStream client and begins the streaming of resources.
+func connectAndStreamResources(logger *zap.SugaredLogger, sm *streamManager) error {
+	resourceCtx, resourceCancel := context.WithCancel(context.Background())
+	defer resourceCancel()
+
+	SendKubernetesResourcesStream, err := sm.streamClient.client.SendKubernetesResources(resourceCtx)
+	if err != nil {
+		logger.Errorw("Failed to connect to server", "error", err)
+		return err
+	}
+
+	sm.streamClient.resourceStream = SendKubernetesResourcesStream
+
+	err = sm.StreamResources(resourceCtx, resourceCancel)
+	if err != nil {
+		logger.Errorw("Failed to bootup and stream resources", "error", err)
+		return err
+	}
+
+	return nil
+}
+
+// connectAndStreamLogs creates sendLogs client and begins the streaming of logs.
+func connectAndStreamLogs(logger *zap.SugaredLogger, sm *streamManager) error {
+	logCtx, logCancel := context.WithCancel(context.Background())
+	defer logCancel()
+
+	SendLogsStream, err := sm.streamClient.client.SendLogs(logCtx)
+	if err != nil {
+		logger.Errorw("Failed to connect to server", "error", err)
+		return err
+	}
+
+	sm.streamClient.logStream = SendLogsStream
+	sm.bufferedGrpcSyncer.UpdateClient(sm.streamClient.logStream, sm.streamClient.conn)
+
+	err = sm.StreamLogs(logCtx, logCancel)
+	if err != nil {
+		logger.Errorw("Failed to bootup and stream logs", "error", err)
+		return err
+	}
+
+	return nil
+}
+
+// Generic function to manage any stream with backoff and reconnection logic.
+func manageStream(logger *zap.SugaredLogger, connectAndStream func(*zap.SugaredLogger, *streamManager) error, sm *streamManager, done chan struct{}) {
 	var backoff = 1 * time.Second
-	authn := Authenticator{Logger: logger}
 	max := big.NewInt(3)
+	rebootCounter := 0
+	// Need a way to prevent operators from all booting at the same time globally if CS has outtage. Maybe a random delay here that gives us some jitter?
 	for {
-		// Generate a random number
-		randomInt, err := rand.Int(rand.Reader, max)
+		err := connectAndStream(logger, sm)
 		if err != nil {
-			logger.Errorw("Could not generate a random int", "error", err)
-			continue
-		}
-		result := randomInt.Int64()
-		sleep := 1*time.Second + backoff + time.Duration(result)*time.Millisecond // Add randomness
-		logger.Infow("Failed to establish connection; will retry", "delay", sleep)
-		time.Sleep(sleep)
-		backoff = backoff * 2 // Exponential increase
-		clientID, clientSecret, err := authn.ReadCredentialsK8sSecrets(ctx, envMap.ClusterCreds)
-		if err != nil {
-			logger.Errorw("Could not read K8s credentials", "error", err)
-		}
-		if clientID == "" && clientSecret == "" {
-			OnboardingCredentials, err := authn.GetOnboardingCredentials(ctx, envMap.OnboardingClientId, envMap.OnboardingClientSecret)
-			if err != nil {
-				logger.Errorw("Failed to get onboarding credentials", "error", err)
-				continue
+			if rebootCounter >= 10 {
+				done <- struct{}{}
```

Comment:
Usually, e.g. like for the `Context.Done()` channel, you should close the channel instead.
```suggestion
				close(done)
```

Right now, nothing is closing the channel, so those channels are leaking.